### ISSUE TITLE
How to use body filter?

### ISSUE THREAD

**Issue by ladjzero** [nginx-clojure/nginx-clojure#114]
Hello Zhang,

Thank you for your awesome module! 
But I got problems while making a custom body filter.
I want to replace all the content from the server and respond the hacked one to clients, playing with nginx proxy_pass. The data is TOTALLY DIFFERENT after being filtered.

There are two filters
```
@Override
    public Object[] doFilter(Map<String, Object> request, InputStream bodyChunk, boolean isLast) throws IOException {
        if (isLast) {
            return new Object[]{200, null, bodyChunk};
        } else {
            return new Object[]{null, null, bodyChunk};
        }
    }
```
```
@Override
    public Object[] doFilter(Map<String, Object> request, InputStream bodyChunk, boolean isLast) throws IOException {
        if (isLast) {
            return new Object[]{200, null, bodyChunk};
        } else {
            return new Object[]{null, null, "hello world"};
        }
    }
```
The first one does not filter anything. The original website and the nginx respond the extremely same thing, as expected.

But the second, just returns status code, and pends forever.

![selection_019](https://cloud.githubusercontent.com/assets/4110311/14454206/cd3e4b40-00cb-11e6-9799-b0e2cb5b81f6.png)

nginx conf
```
location /pipe {
  proxy_set_header X-Real-IP $remote_addr;
  proxy_set_header X-Forwarded-Host $host;
  proxy_set_header X-Forwarded-Server $host;
  proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header X-Forwarded-Proto $scheme;
  proxy_buffering  on;
  proxy_buffer_size 1024k;
  proxy_buffers 4 1024k;
  proxy_pass http://www.hi-pda.com/;
  body_filter_type 'java';
  body_filter_name 'the_first_filter';
}

location /filter {
  proxy_set_header X-Real-IP $remote_addr;
  proxy_set_header X-Forwarded-Host $host;
  proxy_set_header X-Forwarded-Server $host;
  proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header X-Forwarded-Proto $scheme;
  proxy_buffering  on;
  proxy_buffer_size 1024k;
  proxy_buffers 4 1024k;
  proxy_pass http://www.hi-pda.com/;
  body_filter_type 'java';
  body_filter_name 'the_second_filter';
}
```

Transfer-Encoding is chunked. I set a big buffer size to buffer the whole web page.

Did I use body filter in the wrong way? Does body filter satisfy my use case?

**Comment by xfeep** [comment]
@ladjzero Hi, I think it was caused by http header `Content-Length`.
In your case body length was changed to be shorter but http header `Content-Length` was not changed.
Please try add a header filter to remove http header `Content-Length` from response. e.g.

```java
public class MyRemoveContentLengthHeaderFilter implements NginxJavaHeaderFilter {  
      public Object[] doFilter(int status, Map<String, Object> request, Map<String, Object> responseHeaders)
			throws IOException {
		responseHeaders.remove("Content-Length");
                return Constants.PHASE_DONE; 
```

In nginx.conf

```

location /filter {
  .....
 header_filter_type java;
 header_filter_name my.MyRemoveContentLengthHeaderFilter;

}
```

**Comment by ladjzero** [comment]
You commented so soon! Thank you very much.

Thanks to your hint, I found that problem is caused by header `Content-Encoding`. The original website g-zipped the page. But I write strings to client directly without removing the header `Content-Encoding: gzip`.

It works since I removed `Content-Encoding`!

But I got a new question when parsing the input stream in the filter.
Do I need to unzip the input stream by myself on the Java side if the raw data from the server is g-zipped?
